Stop tracking `genReturnBB` after global morph

BruceForstall opened the pull request:
The `genReturnBB` pointer is set if a merged return block is created
during `fgAddInternal` (so, quite early during compilation). It needs
to be maintained (and the unreferenced block it points to needs to be
kept around) until global morph, which hooks up flow to this block.

After global morph, clear the pointer and don't maintain it; it is no
longer needed. Later code that was checking for it can instead check
for `BBJ_RETURN` blocks or `GT_RETURN` nodes. Also, remove the marking
of the `genReturnBB` block as `BBF_DONT_REMOVE`.

This leads to diffs, as unreachable return blocks get deleted. This can
happen, say, if all other exits are tail calls. If that happens and we're
in a case of needing a profiler hook, then the profile exit hook is also
not needed (it would be unreachable).

BruceForstall committed:
Stop tracking `genReturnBB` after global morph

The `genReturnBB` pointer is set if a merged return block is created
during `fgAddInternal` (so, quite early during compilation). It needs
to be maintained (and the unreferenced block it points to needs to be
kept around) until global morph, which hooks up flow to this block.

After global morph, clear the pointer and don't maintain it; it is no
longer needed. Later code that was checking for it can instead check
for `BBJ_RETURN` blocks or `GT_RETURN` nodes. Also, remove the marking
of the `genReturnBB` block as `BBF_DONT_REMOVE`.

This leads to diffs, as unreachable return blocks get deleted. This can
happen, say, if all other exits are tail calls. If that happens and we're
in a case of needing a profiler hook, then the profile exit hook is also
not needed (it would be unreachable).

jakobbotsch reviewed (approved):
LGTM, but seems like win-x86 is unhappy about the removal of (even unreachable) monitor exits in synchronized methods? Presumably in those cases we can just report the entire method (assuming the info is used to exit the lock on exceptional flow by the VM)?

BruceForstall commented:
win-x86 handling of synchronized methods is unique: they aren't transformed to try/finally, but instead are reported to the VM via the GC information and the VM handles unlocking. I'll look into this...

BruceForstall committed:
Allow x86 synchronized methods to have unreachable return blocks

The JIT still needs to report an end-of-synchronized-range offset
in the GC info for x86 synchronized methods, so just create a new
label at the end of all blocks. The VM will automatically exit
the synchronization on an exceptional method exit.

BruceForstall commented:
/azp run runtime-coreclr outerloop, runtime-coreclr jitstress

BruceForstall committed:
Create final IG before clearing GC info

BruceForstall commented:
/azp run runtime-coreclr outerloop, runtime-coreclr jitstress

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=534240&view=ms.vss-build-web.run-extensions-tab)

Pretty significant size improvements.
